Add CEF processor to Ingest node

bhapas opened the pull request:
Closes - https://github.com/elastic/elasticsearch/issues/126201

This PR creates a new CEF ingest node processor. The CEF processor converts a Common Event Format logs into a JSON structure. This processor also maps relevant CEF fields to ECS mappings without a need for additional processors in Ingest pipeline

## Encoding rules from the spec

Ensure the following when encoding symbols in CEF:
- The entire message must be UTF-8 encoded.
- Spaces used in the header are valid. Do not encode a space character by using
`<space>`.
- If a pipe (|) is used in the header, it must be escaped with a backslash (\). But note
that the pipes in the extension do not need escaping. For example:
Sep 19 08:26:10 host CEF:0|security|threatmanager|1.0|100|detected a
\| in message|10|src=10.0.0.1 act=blocked a | dst=1.1.1.1
- If a backslash (\) is used in the header or the extension, it must be escaped with
another backslash (\). For example:
Sep 19 08:26:10 host CEF:0|security|threatmanager|1.0|100|detected a
\\ in packet|10|src=10.0.0.1 act=blocked a \\ dst=1.1.1.1
- If an equal sign (=) is used in the extensions, it has to be escaped with a backslash (\).
Equal signs in the header need no escaping. For example:
Sep 19 08:26:10 host CEF:0|security|threatmanager|1.0|100|detected a =
in message|10|src=10.0.0.1 act=blocked a \= dst=1.1.1.1
- Multi-line fields can be sent by CEF by encoding the newline character as \n or \r.
Note that multiple lines are only allowed in the value part of the extensions. For
example:
Sep 19 08:26:10 host CEF:0|security|threatmanager|1.0|100|Detected a
threat. No action needed.|10|src=10.0.0.1 msg=Detected a threat.\n No
action needed

## Example

An example CEF parsing would look like

<details>
<summary>
CEF LOG
</summary>

```
CEF:0|security|threatmanager|1.0|100|trojan successfully stopped|10|agt=192.168.0.1 agentDnsDomain=example.com ahost=agentHost aid=agentId amac=00:0a:95:9d:68:16 agentNtDomain=example.org art=1622547800000 atz=UTC agentTranslatedAddress=10.0.0.1 agentTranslatedZoneExternalID=ext123 agentTranslatedZoneURI=uri at=agentType av=1.0 agentZoneExternalID=zoneExtId agentZoneURI=zoneUri app=HTTP cnt=1234 in=5678 out=91011 customerExternalID=custExtId customerURI=custUri dst=192.168.0.2 dlat=37.7749 dlong=-122.4194 dhost=destHost dmac=00:0a:95:9d:68:16 dntdom=destNtDomain dpt=80 dpid=1234 dproc=destProc destinationServiceName=destService destinationTranslatedAddress=10.0.0.2 destinationTranslatedPort=8080 destinationTranslatedZoneExternalID=destExtId destinationTranslatedZoneURI=destUri duid=destUserId duser=destUser dpriv=admin destinationZoneExternalID=destZoneExtId destinationZoneURI=destZoneUri act=blocked dvc=192.168.0.3 cfp1Label=cfp1Label cfp3Label=cfp3Label cfp4Label=cfp4Label deviceCustomDate1=1622547800000 deviceCustomDate1Label=customDate1Label deviceCustomDate2=1622547900000 deviceCustomDate2Label=customDate2Label cfp1=1.23 cfp2=2.34 cfp2Label=cfp2Label cfp3=3.45 cfp4=4.56 c6a1=2001:db8::1 c6a1Label=c6a1Label c6a2=2001:db8::2 c6a2Label=c6a2Label c6a3=2001:db8::3 c6a3Label=c6a3Label c6a4=2001:db8::4 C6a4Label=c6a4Label cn1=123 cn1Label=cn1Label cn2=234 cn2Label=cn2Label cn3=345 cn3Label=cn3Label cs1=customString1 cs1Label=cs1Label cs2=customString2 cs2Label=cs2Label cs3=customString3 cs3Label=cs3Label cs4=customString4 cs4Label=cs4Label cs5=customString5 cs5Label=cs5Label cs6=customString6 cs6Label=cs6Label deviceDirection=inbound deviceDnsDomain=example.com cat=category deviceExternalId=extId deviceFacility=16 dvchost=host1 deviceInboundInterface=eth0 dvcmac=00:0a:95:9d:68:16 deviceNtDomain=example.org deviceOutboundInterface=eth1 devicePayloadId=payloadId dvcpid=5678 deviceProcessName=procName rt=1622547800000 dtz=UTC deviceTranslatedAddress=10.0.0.3 deviceTranslatedZoneExternalID=transExtId deviceTranslatedZoneURI=transUri deviceZoneExternalID=zoneExtId deviceZoneURI=zoneUri end=1622547900000 eventId=evt123 outcome=success externalId=extId fileCreateTime=1622547800000 fileHash=abcd1234 fileId=5678 fileModificationTime=1622547900000 fname=file.txt filePath=/path/to/file filePermission=rw-r--r-- fsize=1024 fileType=txt flexDate1=1622547800000 flexDate1Label=flexDate1Label flexString1=flexString1 flexString2=flexString2 flexString1Label=flexString1Label flexString2Label=flexString2Label msg=message oldFileCreateTime=1622547800000 oldFileHash=oldHash oldFileId=oldId oldFileModificationTime=1622547900000 oldFileName=oldFile oldFilePath=/old/path oldFilePermission=rw-r--r-- oldFileSize=2048 oldFileType=oldType rawEvent=rawEvent reason=reason requestClientApplication=Mozilla requestContext=referrer requestCookies=cookies requestMethod=GET request=url src=192.168.0.4 sourceDnsDomain=sourceDomain slat=37.7749 slong=-122.4194 shost=sourceHost smac=00:0a:95:9d:68:16 sntdom=sourceNtDomain spt=443 spid=1234 sproc=sourceProc sourceServiceName=sourceService sourceTranslatedAddress=10.0.0.4 sourceTranslatedPort=8081 sourceTranslatedZoneExternalID=sourceExtId sourceTranslatedZoneURI=sourceUri suid=sourceUserId suser=sourceUser spriv=sourcePriv sourceZoneExternalID=sourceZoneExtId sourceZoneURI=sourceZoneUri start=1622547800000 proto=TCP type=1 catdt=catDeviceType mrt=1622547800000
```
</details>


<details>
<summary>
Parsed CEF content
</summary>

```json
{
  "process": {
    "name": "procName",
    "pid": 5678
  },
  "agent": {
    "ip": "192.168.0.1",
    "name": "example.com",
    "id": "agentId",
    "type": "agentType",
    "version": "1.0",
    "mac": "00:0a:95:9d:68:16"
  },
  "cef": {
    "severity": 10,
    "extensions": {
      "agentTranslatedZoneExternalID": "ext123",
      "flexDate1": "2021-06-01T11:43:20Z",
      "deviceCustomString3Label": "cs3Label",
      "oldFileSize": 2048,
      "sourceZoneURI": "sourceZoneUri",
      "deviceCustomIPv6Address4Label": "c6a4Label",
      "destinationTranslatedZoneURI": "destUri",
      "agentZoneURI": "zoneUri",
      "oldFileName": "oldFile",
      "deviceCustomDate2Label": "customDate2Label",
      "deviceNtDomain": "example.org",
      "deviceCustomFloatingPoint4Label": "cfp4Label",
      "sourceTranslatedZoneURI": "sourceUri",
      "deviceCustomIPv6Address1": "2001:db8::1",
      "deviceCustomDate1Label": "customDate1Label",
      "deviceCustomIPv6Address4": "2001:db8::4",
      "requestCookies": "cookies",
      "deviceCustomIPv6Address3": "2001:db8::3",
      "oldFilePermission": "rw-r--r--",
      "deviceCustomIPv6Address2": "2001:db8::2",
      "deviceCustomString2Label": "cs2Label",
      "deviceCustomFloatingPoint2Label": "cfp2Label",
      "deviceCustomDate2": "2021-06-01T11:45Z",
      "agentTranslatedZoneURI": "uri",
      "deviceCustomDate1": "2021-06-01T11:43:20Z",
      "deviceCustomIPv6Address2Label": "c6a2Label",
      "oldFileModificationTime": "2021-06-01T11:45Z",
      "deviceCustomFloatingPoint1": 1.23,
      "oldFileHash": "oldHash",
      "deviceCustomFloatingPoint2": 2.34,
      "deviceCustomFloatingPoint3": 3.45,
      "flexString1": "flexString1",
      "deviceCustomFloatingPoint4": 4.56,
      "oldFileId": "oldId",
      "deviceCustomNumber1": 123,
      "agentTranslatedAddress": "10.0.0.1",
      "deviceCustomNumber3": 345,
      "deviceCustomNumber2": 234,
      "flexString2": "flexString2",
      "baseEventCount": 1234,
      "deviceCustomIPv6Address1Label": "c6a1Label",
      "deviceTranslatedZoneExternalID": "transExtId",
      "deviceZoneExternalID": "zoneExtId",
      "agentTimeZone": "UTC",
      "deviceCustomString6Label": "cs6Label",
      "deviceCustomNumber2Label": "cn2Label",
      "deviceCustomString5Label": "cs5Label",
      "deviceCustomFloatingPoint1Label": "cfp1Label",
      "sourceZoneExternalID": "sourceZoneExtId",
      "deviceTranslatedZoneURI": "transUri",
      "destinationTranslatedZoneExternalID": "destExtId",
      "flexString1Label": "flexString1Label",
      "deviceCustomNumber1Label": "cn1Label",
      "categoryDeviceType": "catDeviceType",
      "deviceZoneURI": "zoneUri",
      "flexString2Label": "flexString2Label",
      "deviceCustomNumber3Label": "cn3Label",
      "deviceCustomString1": "customString1",
      "externalId": "extId",
      "oldFilePath": "/old/path",
      "deviceCustomString3": "customString3",
      "deviceCustomString2": "customString2",
      "deviceCustomString1Label": "cs1Label",
      "deviceCustomString5": "customString5",
      "deviceCustomString4": "customString4",
      "agentZoneExternalID": "zoneExtId",
      "oldFileCreateTime": "2021-06-01T11:43:20Z",
      "deviceCustomString6": "customString6",
      "deviceCustomIPv6Address3Label": "c6a3Label",
      "deviceEventCategory": "category",
      "deviceCustomString4Label": "cs4Label",
      "deviceCustomFloatingPoint3Label": "cfp3Label",
      "destinationZoneExternalID": "destZoneExtId",
      "flexDate1Label": "flexDate1Label",
      "sourceTranslatedZoneExternalID": "sourceExtId",
      "agentNtDomain": "example.org",
      "oldFileType": "oldType",
      "destinationZoneURI": "destZoneUri"
    },
    "device.version": "1.0",
    "name": "trojan successfully stopped",
    "device.vendor": "security",
    "device.product": "threatmanager",
    "device.event_class_id": 100,
    "version": 0
  },
  "log": {
    "syslog": {
      "facility": {
        "code": 16
      }
    }
  },
  "destination": {
    "nat": {
      "port": 8080,
      "ip": "10.0.0.2"
    },
    "geo": {
      "location": {
        "lon": -122.4194,
        "lat": 37.7749
      }
    },
    "registered_domain": "destNtDomain",
    "process": {
      "name": "destProc",
      "pid": 1234
    },
    "port": 80,
    "bytes": 91011,
    "service": {
      "name": "destService"
    },
    "domain": "destHost",
    "ip": "192.168.0.2",
    "user": {
      "name": "destUser",
      "id": "destUserId",
      "group": {
        "name": "admin"
      }
    },
    "mac": "00:0a:95:9d:68:16"
  },
  "source": {
    "geo": {
      "location": {
        "lon": -122.4194,
        "lat": 37.7749
      }
    },
    "nat": {
      "port": 8081,
      "ip": "10.0.0.4"
    },
    "registered_domain": "sourceNtDomain",
    "process": {
      "name": "sourceProc",
      "pid": 1234
    },
    "port": 443,
    "service": {
      "name": "sourceService"
    },
    "bytes": 5678,
    "ip": "192.168.0.4",
    "domain": "sourceDomain",
    "user": {
      "name": "sourceUser",
      "id": "sourceUserId",
      "group": {
        "name": "sourcePriv"
      }
    },
    "mac": "00:0a:95:9d:68:16"
  },
  "message": "message",
  "url": {
    "original": "url"
  },
  "network": {
    "protocol": "HTTP",
    "transport": "TCP",
    "direction": "inbound"
  },
  "observer": {
    "ingress": {
      "interface": {
        "name": "eth0"
      }
    },
    "registered_domain": "example.com",
    "product": "threatmanager",
    "hostname": "host1",
    "vendor": "security",
    "ip": "192.168.0.3",
    "name": "extId",
    "version": "1.0",
    "mac": "00:0a:95:9d:68:16",
    "egress": {
      "interface": {
        "name": "eth1"
      }
    }
  },
  "file": {
    "inode": 5678,
    "path": "/path/to/file",
    "size": 1024,
    "created": "2021-06-01T11:43:20Z",
    "name": "file.txt",
    "mtime": "2021-06-01T11:45Z",
    "type": "txt",
    "hash": "abcd1234",
    "group": "rw-r--r--"
  },
  "@timestamp": "2021-06-01T11:43:20Z",
  "organization": {
    "name": "custUri",
    "id": "custExtId"
  },
  "host": {
    "nat": {
      "ip": "10.0.0.3"
    }
  },
  "http": {
    "request": {
      "referrer": "referrer",
      "method": "GET"
    }
  },
  "event": {
    "reason": "reason",
    "ingested": "2021-06-01T11:43:20Z",
    "original": "rawEvent",
    "code": 100,
    "kind": 1,
    "created": "2021-06-01T11:43:20Z",
    "timezone": "UTC",
    "start": "2021-06-01T11:43:20Z",
    "action": "blocked",
    "end": "2021-06-01T11:45Z",
    "id": "evt123",
    "outcome": "success"
  },
  "user_agent": {
    "original": "Mozilla"
  }
}
```
</details>

<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

---

- [x] Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- [x] Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md)?
- [ ] If submitting code, have you built your formula locally prior to submission with `gradle check`?
- [x] If submitting code, is your pull request against main? Unless there is a good reason otherwise, we prefer pull requests against main and will backport as needed.
- [x] If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- [ ] If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

bhapas committed:
Marius base code

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
add cef processor test

bhapas committed:
Merge branch 'cef_processor' of github.com:bhapas/elasticsearch into cef_processor

andrewkroh commented:
I realize this draft is still in progress, and you likely already have plans for these items.

1. Once this merges, the [elastic/elasticsearch-specification](https://github.com/elastic/elasticsearch-specification/blob/265cd5572fbfdce7c16dd5c151d57ac2010624d1/specification/ingest/_types/Processors.ts#L32) will need to be updated to include the new processor and all of its parameters.
2. The existing test suites for the `decode_cef` processor in Beats should be incorporated here. If there are any behavioral differences, we should identify them and evaluate whether they are justified. Our goal is to ensure that usages of `decode_cef` can be replaced with this new processor seamlessly.
3. While adherence to the specification is the priority, it would also be interesting to compare performance metrics between the new processor and the existing `decode_cef` processor. The latter has [two microbenchmarks](https://github.com/elastic/beats/blob/e4012a23dc45ba15695dae56489c7087c0eb7797/x-pack/filebeat/processors/decode_cef/decode_cef_test.go#L381-L402), one for short messages and another for long messages. If we have the capability to conduct microbenchmarks, let's add these same two tests.

Additionally, there is a CEF v1 specification (our `decode_cef` processor was based on CEF v0). It would be worthwhile to review if any major changes in CEF v1 could impact the architecture of our code. See [this issue](https://github.com/elastic/beats/issues/40269) for more details.

bhapas committed:
Modify the parser and add more tests

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
Fix parser and add more tests

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
Move ecs mapping to root

bhapas committed:
Merge branch 'cef_processor' of github.com:bhapas/elasticsearch into cef_processor

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
move non-ecs fields under cef.extensions

bhapas committed:
Merge branch 'cef_processor' of github.com:bhapas/elasticsearch into cef_processor

Commit:
[CI] Auto commit changes from spotless

andrewkroh commented:
I asked Lee H about micro-benchmarking, and [JMH](https://github.com/openjdk/jmh) is being used (see https://github.com/elastic/elasticsearch/tree/main/benchmarks#elasticsearch-microbenchmark-suite). So this could add a benchmark under that suite of tests.

bhapas commented:
> I asked Lee H about micro-benchmarking, and [JMH](https://github.com/openjdk/jmh) is being used (see https://github.com/elastic/elasticsearch/tree/main/benchmarks#elasticsearch-microbenchmark-suite). So this could add a benchmark under that suite of tests.

Will this be comparable to the microbenchmarking that is done in the beats processor?

bhapas committed:
move ecs fields to ingestdocument root

bhapas committed:
Merge branch 'main' into cef_processor

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
fix tests

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
Fix commit issues

bhapas committed:
fix negative test

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,449 @@
+        return extensions;
+    }
+
+    private void removeEmptyValue(Map<String, String> map) {
```
Methods that do not need access to instance members should be marked as `static`.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,449 @@
+
+        public void removeMappedExtensions() {
+            this.extensions.entrySet().removeIf(entry -> FIELD_MAPPING.containsKey(entry.getKey()));
+            this.extensions.entrySet().removeIf(entry -> FIELD_MAPPING.containsValue(entry.getKey()));
```
You're doing N scans through the whole collection here, I doubt that's going to be especially performant.

bhapas replied:
I think this can be 

```java
public void removeMappedExtensions() {
    this.extensions.entrySet().removeIf(entry -> 
        FIELD_MAPPING.containsKey(entry.getKey()) || FIELD_MAPPING.containsValue(entry.getKey())
    );
}
```

Or you mean something else?

joegallo replied:
Explain to me how `containsValue` works on a Map.

bhapas replied:
Ohh yeah I get your point now.. I was only looking at `containsKey` thing. Yeah the `containsValue` iterates over the   N entries each time. I ll figure out a different way for this.

joegallo replied:
If you smash all the keys and all the values into a single additional static `Set`, then you could do a `this.extensions.removeAll(thatStaticSet)`.

bhapas replied:
```
        public void removeMappedExtensions() {
            Set<String> fieldMappingKeysAndValues = new HashSet<>();
            fieldMappingKeysAndValues.addAll(FIELD_MAPPING.keySet());
            fieldMappingKeysAndValues.addAll(FIELD_MAPPING.values());

            extensions.entrySet().removeIf(entry -> fieldMappingKeysAndValues.contains(entry.getKey()));
        }
```

bhapas replied:
Moved to static Hashset.

joegallo replied:
This looks right to me now, but I moved the location of the definition of the static set.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,449 @@
+        } else if (type == Integer.class) {
+            return Integer.parseInt(value);
+        } else {
+            return value; // Default to String
```
If the default is `String`, then I'm not sure I see the benefit of the many explicit String-typed mappings in `ECSFieldWithType`.

joegallo replied:
I added 5c3ee2ba86155cd6bf6a5eb678f7c1603de3dcf4 and now I'm resolving this.

joegallo commented:
Overall this is pretty neat and I like the core of it. I think there's a version of it that's a bit tidier inside that we can all find together, but +1 on the general notion and fundamental direction of this. I think this'll be a nice additional capability for Elasticsearch to have.

bhapas committed:
Update date time handling

bhapas commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,449 @@
+        }
+    }
+
+    private static class ECSFieldWithType {
```
This class is removed in favour of static mapping

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,449 @@
+    private static final Pattern EXTENSION_NEXT_KEY_VALUE_PATTERN = Pattern.compile(
+        "(" + EXTENSION_KEY_PATTERN + ")=(" + EXTENSION_VALUE_PATTERN + ")(?:\\s+|$)"
+    );
+    private static final Map<String, String> EXTENSION_VALUE_SANITIZER_REVERSE_MAPPING = new HashMap<>();
```
I don't see this change on the PR, so I unresolved this conversation.

bhapas replied:
Fixed in [4a1a9ac](https://github.com/elastic/elasticsearch/pull/122491/commits/4a1a9ac5164bd3ec314e392b440ac84e47faafa0)

joegallo replied:
There's more that can be done. You don't need a static initializer in this class, all your collections can be constructed via the static collection factory methods (`Set.of`, `Set.copyOf`, ..., `Map.ofEntries`, etc). Then you won't have static mutable maps in this class anymore.

joegallo replied:
This looks right to me now.

bhapas committed:
fix review comments

bhapas committed:
Merge branch 'main' into cef_processor

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -91,7 +104,7 @@ final class CefParser {
     );
 
-    private final List<ChronoField> FIELDS = Arrays.asList(
+    private final List<ChronoField> CHRONO_FIELDS = Arrays.asList(
```
`List.of` is fewer characters than `Arrays.asList`.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -177,6 +185,80 @@ final class CefParser {
+        FIELDS.put("url.original", String.class);
+        FIELDS.put("user_agent.original", String.class);
+
+        FILED_MAPPINGS_AND_VALUES.addAll(FIELD_MAPPING.keySet());
```
Typo in variable name.

bhapas replied:
Nice catch

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -101,11 +114,6 @@ final class CefParser {
     );
 
     static {
```
Drop the static initializer and use `Map.ofEntries`.

bhapas replied:
Dropped

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -61,8 +62,20 @@ final class CefParser {
         "(" + EXTENSION_KEY_PATTERN + ")=(" + EXTENSION_VALUE_PATTERN + ")(?:\\s+|$)"
     );
-    private static final Map<String, String> EXTENSION_VALUE_SANITIZER_REVERSE_MAPPING = new HashMap<>();
+    private static final Map<String, String> EXTENSION_VALUE_SANITIZER_REVERSE_MAPPING = Map.of(
```
Use `Map.ofEntries` here so it's readable. Like this:

```diff
+++ b/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java
@@ -39,6 +39,7 @@ import static java.time.temporal.ChronoField.MINUTE_OF_DAY;
 import static java.time.temporal.ChronoField.MONTH_OF_YEAR;
 import static java.time.temporal.ChronoField.NANO_OF_SECOND;
 import static java.time.temporal.ChronoField.SECOND_OF_DAY;
+import static java.util.Map.entry;

 final class CefParser {
     private final IngestDocument ingestDocument;
@@ -62,15 +63,11 @@ final class CefParser {
     private static final Pattern EXTENSION_NEXT_KEY_VALUE_PATTERN = Pattern.compile(
         "(" + EXTENSION_KEY_PATTERN + ")=(" + EXTENSION_VALUE_PATTERN + ")(?:\\s+|$)"
     );
-    private static final Map<String, String> EXTENSION_VALUE_SANITIZER_REVERSE_MAPPING = Map.of(
-        "\\\\",
-        "\\",
-        "\\=",
-        "=",
-        "\\\n",
-        "\n",
-        "\\\r",
-        "\r"
+    private static final Map<String, String> EXTENSION_VALUE_SANITIZER_REVERSE_MAPPING = Map.ofEntries(
+        entry("\\\\", "\\"),
+        entry("\\=", "="),
+        entry("\\\n", "\n"),
+        entry("\\\r", "\r)")
     );
```

bhapas committed:
fix more comments

bhapas committed:
Make parser independent of ingestDocument

bhapas commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,449 @@
+        return desanitized;
+    }
+
+    public static class CEFEvent {
```
Modified this to just hold `rootMappings` and `cefMappings` Maps.

joegallo replied:
Nice, I like that a lot!

bhapas commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,449 @@
+
+final class CefParser {
+
+    private final IngestDocument ingestDocument;
```
Removed `ingestDocument` from Parser. Instead parser creates a `CEFEvent` object and returns to Processor where it is painted onto the ingestDocument. Fixed in https://github.com/elastic/elasticsearch/pull/122491/commits/0008f8ad85434469ca8bf06cad5430a60c5ce771

joegallo replied:
YES! I like that a lot better. Very nice.

bhapas committed:
Pass copy of headerFields to process

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
return fast if type is string

joegallo committed:
Prefer List.of over Arrays.asList

joegallo committed:
Rename this variable

joegallo committed:
Move this definition

so that it's closer to the other static member from which it derives

joegallo committed:
Do not copy these maps

joegallo committed:
Conserve precious characters

joegallo committed:
Whitespace

joegallo committed:
Silence some warnings from IntelliJ

joegallo committed:
Allocate the immutable set statically

joegallo committed:
There's no need to copy this list

joegallo committed:
Conserve precious characters

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,459 @@
+        }
+
+        // Translate extensions to possible ECS fields
+        Map<String, Object> translatedFields = extensions.entrySet()
```
Please rewrite this streaming block into ordinary banging on collections -- this is code is too hot for streaming (also there are some wasteful things you're doing in the streaming version that will drop out trivially in the non-streaming version.

bhapas committed:
Implement MACAddress and IP types

bhapas committed:
Change to traditional looping

bhapas committed:
Merge branch 'main' into cef_processor

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -348,13 +347,15 @@ private void processExtensions(String cefString, int extensionStart, CEFEvent ev
-            }));
+        Map<String, Object> translatedFields = new HashMap<>();
+        for (Map.Entry<String, String> entry : extensions.entrySet()) {
+            if (FIELD_MAPPINGS.containsKey(entry.getKey())) {
```
Don't do a `containsKey` and then a `get`.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -348,13 +347,15 @@ private void processExtensions(String cefString, int extensionStart, CEFEvent ev
-                DataType fieldType = FIELDS.get(FIELD_MAPPINGS.get(entry.getKey()));
-                return convertValueToType(entry.getValue(), fieldType);
-            }));
+        Map<String, Object> translatedFields = new HashMap<>();
```
Do you think it's worth pre-sizing this `HashMap` (via `HashMap.newHashMap`) to match the size of the `extensions`? By my read that would slightly oversize the map, but with the effect that you'd never need to resize. I imagine it's worth it.

bhapas replied:
Removing all intermediate collections. Adding the mapping to event as we parse.

bhapas committed:
Handle missed keys for different CEF logs

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Merge branch 'main' into cef_processor

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -516,22 +536,53 @@ private static String desanitizeExtensionVal(String value) {
         }
     }
+
+    public static class ExtensionMapping {
```
Use a java record.

bhapas replied:
Fixed in https://github.com/elastic/elasticsearch/pull/122491/commits/494a31b4c304fe51de179c3ce1b63d89676c9c30

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -516,22 +536,53 @@ private static String desanitizeExtensionVal(String value) {
 
-        public Map<String, String> getCefMappings() {
+        public Map<String, Object> getCefMappings() {
+            cefMappings.put("extensions", extensions);
```
Something is very wrong if this is the right solution here.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -392,12 +413,12 @@ private static Map<String, String> parseExtensions(String extensionString) {
     }
 
     private Object convertValueToType(String value, DataType type) {
```
The body of this method would probably be better if it were implemented as a big switch/case rather than a long chain of if/elses.

bhapas replied:
https://github.com/elastic/elasticsearch/pull/122491/commits/7ee0213f357cca149bdb0ddc2397dcd0fd5b85d6

joegallo committed:
Conserve precious characters

joegallo commented:
You'll use way fewer characters (and gain a bunch of **clarity**) in `CefProcessorTests` if you stop using `getFieldValue` line by line (that's not really what it's for, anyway) and instead use `getSourceAndMetadata()` to grab a map of the underlying details just once.

bhapas committed:
do not modify collection in getter

bhapas committed:
convert if/else to switch

joegallo committed:
Don't use a default here

Then the compiler will error out in the future if a new datatype is
added to the enum but not added to this switch.

joegallo committed:
Reorder these conditions

bhapas committed:
Add extensions only if not empty

bhapas committed:
Use record for ExtensionMapping

bhapas committed:
private method

bhapas committed:
fix test to match source than each field

joegallo committed:
Use Map.of for small Maps

joegallo committed:
Drop an unnecessary annotation

joegallo committed:
Rename this variable

joegallo committed:
Handle the extensions with processExtensions

joegallo committed:
Drop extensions from CEFEvent entirely

joegallo committed:
Manage the event lifecycle a little more tightly

This prevents us from needing to return a copy or from needing to put
the maps in an unmodifiable wrapper. The object is single use: you
build it, get the maps, close it, and then you're done.

joegallo committed:
Fuss with the visibility

joegallo committed:
Nest the cef mappings correctly as sub-objects

joegallo committed:
Merge branch 'main' into cef_processor

bhapas committed:
verify tests against the whole document source

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
Do not allow escaped pipe in extensions

joegallo committed:
Use a statement lambda

which silences a warning from IntelliJ

joegallo committed:
Prefer Map.of when it's a one-liner

joegallo committed:
There's no need to have these as variables

joegallo committed:
Add a utility for comparing maps

So that when comparing large maps it's clearer where the difference
between the two maps actually is.

bhapas committed:
Change exception message

bhapas committed:
Merge branch 'main' into cef_processor

joegallo committed:
One of these things is not like the others

joegallo committed:
This can be static

joegallo committed:
Conserve precious characters

joegallo committed:
Flatten the map structure

joegallo committed:
Tighten up this implementation

joegallo committed:
Whitespace

joegallo committed:
Rename this parameter and some variables

joegallo committed:
Conserve precious characters

joegallo committed:
Reorder these statements

joegallo commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/CefProcessorTests.java:
```diff
@@ -0,0 +1,1104 @@
+        assertEquals("Invalid IP address format", exception.getMessage());
+    }
+
+    public void toIP_nullString() {
```
These `toIP_` methods aren't formatted to actually be JUnit tests (which start with `test`), so none of them are running. If you fix that so that they do run, not all of the tests pass.

Please avoid underscores in these names, regardless, so for example have `toIP_nullString` become `testToIPNullString`.

bhapas replied:
https://github.com/elastic/elasticsearch/pull/122491/commits/2b5bfd2ce93a8a8e1d043a6fd7f073585f70945b fixes this

joegallo committed:
The dataType is never null

bhapas committed:
Fix test naming

The toIp tests with null / empty string are still valid
Referring to docs from InetAddress.getByName()
"If the host is null or host.length() is equal to zero,
then an InetAddress representing an address of the loopback interface is returned."

bhapas committed:
Merge branch 'main' into cef_processor

joegallo committed:
Use the InetAddresses utility for this

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,564 @@
+        // Insert separators if necessary
+        String macWithSeparators = insertMACSeparators(v);
+        // Validate MAC address format
+        Pattern macAddressPattern = Pattern.compile(MAC_ADDRESS_REGEX);
```
Don't compile regexes at runtime in this way -- do it statically.

bhapas replied:
https://github.com/elastic/elasticsearch/pull/122491/commits/9aa97a16053a96989ece51ac5eaf47afe39dbd9d

bhapas committed:
make private

bhapas committed:
Make Pattern compilation static

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Merge branch 'main' into cef_processor

joegallo commented:
Is it the intention that the timezone for a date can be specified as a `String` (**only**) or is it necessary that a timezone can be specified as a templated-expression that allows a timezone to be dynamically looked up in some other part of the incoming document?

The [beats version of things](https://www.elastic.co/guide/en/beats/filebeat/current/processor-decode-cef.html) appears to only allow a `String`, but this current implementation allows a templated-expression.

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,565 @@
+            CefEvent event = new CefEvent();
+            // Add error message if there are not enough header fields
+            if (headers.size() != 7) {
+                event.addRootMapping("error.message", new HashSet<>(ERROR_MESSAGE_INCOMPLETE_CEF_HEADER));
```
Why does this implementation add an error message and not throw an exception? This doesn't seem in line with the way ingest processing usually handles errors.

bhapas replied:
I followed beats approach for this initially. But I agree that this should just throw exception just like other scenarios.

bhapas commented:
> Is it the intention that the timezone for a date can be specified as a `String` (**only**) or is it necessary that a timezone can be specified as a templated-expression that allows a timezone to be dynamically looked up in some other part of the incoming document?

The intention is that the timezone in this processor can be a templated expression since it can already be defined in other field. I think this way of looking up timezone dynamically is there in other ingest node processors too. I can lookup to see why this was ignored in beats processor. But it is intentional to allow dynamic templates to fill in timezone

bhapas committed:
Throw exception for incomplete CEF header

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,565 @@
+    );
+
+    private static final Set<String> ERROR_MESSAGE_INCOMPLETE_CEF_HEADER = Set.of("incomplete CEF header");
+    private static final List<String> TIME_LAYOUTS = List.of(
```
Let's do better than that. How about comment with a link to some kind of authoritative source, and let's review these formats in terms of that authoritative source?

bhapas replied:
https://github.com/elastic/elasticsearch/pull/122491/commits/4470e10ee48f226b5ad40cf19ec3c3097fba7ed3 adds a doc note referring the official documentation

joegallo replied:
I'm not sure the hardcoded `'GMT'` strings in the list below are supported by the documentation you linked. Do these format strings all actually make sense in the context of the linked reference material?

joegallo replied:
Heh, I would note for the record that the spec appears to be wrong to me, though. There's a typo in the `MMM dd HH:mm:sss.SSS` format string there, it seems natural that what was actually intended by the spec would be `MMM dd HH:mm:ss.SSS`.

joegallo replied:
2fecc52a9008c6d7e82f25c350ec719119e1da4f seems to pass the tests to me, and has the benefit of matching the list of date formats that we use to the list of date formats from the spec.

bhapas committed:
Add doc for Timestamp formats

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Add appendix details

bhapas committed:
Merge branch 'main' into cef_processor

joegallo committed:
Simplify the list of supported formats

The list of formats is exactly those supported by the spec, no more no
less (modulo that there's a typo in the spec that I've corrected
here).

joegallo commented:
It seems to me like your handling of the `deviceFacility` is wrong according to the spec.

joegallo reviewed (changes requested):
It seems to me like your handling of the `deviceFacility` is wrong according to the spec.

joegallo commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/CefProcessorTests.java:
```diff
@@ -0,0 +1,1065 @@
+                entry("event", Map.of("code", "100")),
+                entry("observer", Map.of("product", "threatmanager", "vendor", "security", "version", "1.0")),
+                entry("source", Map.of("ip", "10.0.0.192", "port", 1232)),
+                entry("message", "Trailing space in non-final extensions is  preserved")
```
```
joegallo@simulacron:~/Code/elastic/beats $ git grep "is  preserved"
x-pack/filebeat/processors/decode_cef/cef/cef_test.go:  paddedMessage = `CEF:0|security|threatmanager|1.0|100|message is padded|10|spt=1232 msg=Trailing space in non-final extensions is  preserved    src=10.0.0.192 `
x-pack/filebeat/processors/decode_cef/cef/cef_test.go:                  "msg": StringField("Trailing space in non-final extensions is  preserved   "),
```

It seems to like this test is written incorrectly, and that the behavior here is wrong. The beats version of this seems to have to the correct idea.

bhapas replied:
Handled https://github.com/elastic/elasticsearch/pull/122491/commits/3f7c3d468b6f26724e4cc173cb29f1ef6beda435

joegallo committed:
Rewrite the header processing to avoid regexes

This uses constant stack and it's about 20% faster overall for the
entire processor. That latter bit is impressive considering that the
vast majority of the time spent in the CefParser is spent in
processExtensions (the other regex-y stuff that I didn't touch).

joegallo committed:
Merge branch 'main' into cef_processor

joegallo commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,561 @@
+        map.values().removeIf(Strings::isEmpty);
+    }
+
+    private static String convertArrayLikeKey(String key) {
```
I don't think `convertArrayLikeKey` or the path that ends up calling it is actually covered by any tests. Ideally, I would like to see a document from the real world that exercises this code path.

bhapas replied:
Removed this code as it 
 is not used https://github.com/elastic/elasticsearch/pull/122491/commits/a3bd7ab66f4d3f70cd47ce8e9f0073c43d5872a9

bhapas committed:
Fix datatypes as per spec
The field datatypes are fixed according the CEF spec V27

bhapas committed:
Fix reason to lowercase

bhapas committed:
Change label to lowercase

bhapas committed:
Change deviceFacility to String per spec

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Correct field types according to spec

bhapas committed:
Add CEF spec 1.x fields

bhapas commented:
> c71e855

Handled all the fields according to the latest spec

bhapas committed:
Remove unnnecessary code

bhapas committed:
Remove unused constant

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Fix handling of spaces

bhapas commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/CefProcessorTests.java:
```diff
@@ -0,0 +1,1065 @@
+                entry("event", Map.of("code", "100")),
+                entry("observer", Map.of("product", "threatmanager", "vendor", "security", "version", "1.0")),
+                entry("source", Map.of("port", 1232, "ip", "127.0.0.1")),
+                entry("message", "Tabs\tand\rcontrol\ncharacters are preserved")
```
Handled here https://github.com/elastic/elasticsearch/pull/122491/commits/3f7c3d468b6f26724e4cc173cb29f1ef6beda435

bhapas committed:
Throw exception on unescaped equals sign
Unescaped equals sign in extension values should be rejected.
CEF v27 spec as well as beats impl does this

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Merge branch 'main' into cef_processor

andrewkroh commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java:
```diff
@@ -0,0 +1,601 @@
+
+    /**
+     * List of allowed timestamp formats for CEF spec v27, see: Appendix A: Date Formats
+     * <a href="https://www.microfocus.com/documentation/arcsight/arcsight-smartconnectors-24.2/pdfdoc/cef-implementation-standard/cef-implementation-standard.pdf">documentation</a>
```
I recommend using an archive.org link to help the references stand the test of time. Plus it ensures the integrity of the content such that the v27 reference material does not change on us.

https://web.archive.org/web/20240925170528/http://www.microfocus.com/documentation/arcsight/arcsight-smartconnectors-24.2/pdfdoc/cef-implementation-standard/cef-implementation-standard.pdf

bhapas replied:
https://github.com/elastic/elasticsearch/pull/122491/commits/8f332391fa22aaa52c540874ea45de44afb5dda2

andrewkroh commented:
> The behavioral differences are related to the CEF V1 spec which are added additionally to this processor.

Is https://github.com/elastic/beats/issues/40269#issuecomment-2800604004 a full accounting of the CEF v1 changes accounted for? If not, please update that comment to be comprehensive because I would like to eventually have the equivalent CEF v1 handling in the beats code too. Thanks.

andrewkroh reviewed (commented):
Nice work! Thank you for the guidance, Joe.

Do we need to include the documentation changes in this pull request? My preference is to include them alongside the code changes, but I'm not sure if Elasticsearch follows a different process. If they are separate pull requests, you might consider adding a usage example (using `POST _ingest/pipeline/_simulate`) and a configuration option table in the PR description.

bhapas commented:
> > The behavioral differences are related to the CEF V1 spec which are added additionally to this processor.
> 
> Is [elastic/beats#40269 (comment)](https://github.com/elastic/beats/issues/40269#issuecomment-2800604004) a full accounting of the CEF v1 changes? If not, could you please update that comment to ensure it's comprehensive? I would like to eventually implement equivalent CEF v1 handling in the Beats code. Thanks.

@andrewkroh The comment depicts changes for fields that are already present in V0 that changed in V1. There are few fields that are introduced in V1. I will add them too

joegallo committed:
Merge branch 'main' into cef_processor

joegallo committed:
Clean the mac address compilation up a little

joegallo committed:
Remove these unused types

joegallo commented:
@bhapas are there any benchmarks associated with the beats implementation of this parser? It would be interesting to compare performance of the two implementations.

bhapas commented:
> 3\. The latter has [two microbenchmarks](https://github.com/elastic/beats/blob/e4012a23dc45ba15695dae56489c7087c0eb7797/x-pack/filebeat/processors/decode_cef/decode_cef_test.go#L381-L402), one for short messages and another for long messages

@joegallo Yes beats has [two microbenchmarks](https://github.com/elastic/beats/blob/e4012a23dc45ba15695dae56489c7087c0eb7797/x-pack/filebeat/processors/decode_cef/decode_cef_test.go#L381-L402), one for short messages and another for long messages

joegallo commented:
In the same vein, do we have any kind of order-of-magnitude for how many documents we can expect to pass through this processor in the real world? My attention to performance will be laxer if this parser is going to see several hundred documents per hour than if it's going to receive several thousand documents per second.

joegallo committed:
Tweak this method signature

joegallo committed:
Rework these throw blocks into early returns

joegallo committed:
Add a default condition, just in case

joegallo committed:
Separate parsing+processing for headers and extensions

joegallo committed:
Drop this setUp method

joegallo committed:
Add a (relatively) simple stub test of the parsing

joegallo committed:
Merge branch 'main' into cef_processor

joegallo committed:
Manage a static list of DateTimeFormatters

rather than re-creating them on each invocation.

joegallo committed:
Move this enum to the bottom

so it's near ExtensionMapping

joegallo committed:
Drop the pattern list constant

joegallo committed:
Add a test that should fail

joegallo committed:
Make the test fail

Pipe should not be a valid separator in mac addresses

joegallo committed:
Standardize on expectThrows

joegallo committed:
Rewrite the mac address regexing

- there's no need for the beginning/end match (that's implied as far as I can tell)
- drop some unnecessary groups
- use non-capturing groups since we're just looking for a match
- as a consequence this can be constructed with String.join
- add a few more tests

joegallo committed:
Micro-optimization: just use a new buffer

GCing old buffers and allocating new ones is very fast, reseting the
buffer requires rewriting the backing array in a way that's slow
enough to show up on a flamegraph.

joegallo committed:
Add a couple more test cases

joegallo committed:
Add a new utility method for trimming strings

bhapas commented:
> @bhapas can you take a look at [6f10616](https://github.com/elastic/elasticsearch/commit/6f106163cc3fa822ac5aad518d4ba2f38b0ed548) (and [4d03454](https://github.com/elastic/elasticsearch/commit/4d0345431f9eac554185cedb0ae4bc21041ccee1))?

@joegallo Looks good.

Just a little confusion. Why are https://github.com/elastic/elasticsearch/pull/122491/commits/45a1af3db831af4f43c3364535a25dfdd58cadb5 and https://github.com/elastic/elasticsearch/pull/122491/commits/65a91ae885f4e726644da22446a13da05db660e1 part of this PR? Are they relevant for any further changes intended, unless I am blind?

bhapas committed:
Use web.archive.org link

bhapas commented:
>Do we need to include the documentation changes in this pull request? My preference is to include them alongside the code changes, but I'm not sure if Elasticsearch follows a different process

@joegallo Your opinion on this please? cc: @andrewkroh

bhapas committed:
Merge branch 'main' into cef_processor

joegallo commented:
> Just a little confusion. Why are https://github.com/elastic/elasticsearch/commit/45a1af3db831af4f43c3364535a25dfdd58cadb5 and https://github.com/elastic/elasticsearch/commit/65a91ae885f4e726644da22446a13da05db660e1 part of this PR? Are they relevant for any further changes intended, unless I am blind?

I have WIP that I haven't pushed yet.

joegallo committed:
Collapse the time formats into a single format

joegallo committed:
Merge branch 'main' into cef_processor

joegallo committed:
Inline ignoring empty values

bhapas committed:
Merge branch 'main' into cef_processor

joegallo reviewed (changes requested):
My read of this is that you are not following the spec for newlines and carriage returns. (edit: I've elaborated in https://github.com/elastic/elasticsearch/pull/122491#pullrequestreview-3111408630).

joegallo commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/CefProcessorTests.java:
```diff
@@ -0,0 +1,1040 @@
+
+    public void testEscapedMessage() {
+        String message = "CEF:0|security\\compliance|threat\\|->manager|1.0|100|message contains escapes|10|"
+            + "spt=1232 msg=Newlines in messages\\\nare allowed.\\\r\\\nAnd so are carriage feeds\\\\newlines\\\\\\=. dpt=4432";
```
This string in java contains the following substrings, but I'm going to mentally unescape the java string escaping rules:

`messages{BACKSLASH}{CARRIAGERETURN}are allowed`

`allowed.{BACKSLASH}{LINEFEED}{BACKSLASH}{CARRIAGERETURN}And`

That's not the escaping that's written into the spec. My read of the spec is that the existence of actual carriage return and linefeed in the input is entirely irrelevant (indeed perhaps non-existent in well formed inputs?) -- rather, the spec says that if you see an actual backlash character followed by an actual `n` or `r` character then those need to be **treated** as being carriage returns and linefeeds.


So here the test and the code are checking the wrong thing, or at least that's my read of the code and the java escaping and the cef escaping and the spec.

It's entirely possible that the existing beats implementation has these same bugs, I haven't checked.

bhapas committed:
Merge remote-tracking branch 'upstream/main' into cef_processor

bhapas committed:
Move cef messages to text ficture files

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
change to .cef.txt to avoid ruling global .txt files out

bhapas committed:
Merge remote-tracking branch 'upstream/main' into cef_processor

bhapas committed:
more test fixtures migration

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
Move to test files

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
Fix backslash handling with newline , tabs , carriagefeeds

Commit:
[CI] Auto commit changes from spotless

bhapas committed:
Protect Escaped backslash

bhapas committed:
Fix test with tab as a separator

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Merge branch 'main' into cef_processor

bhapas committed:
Use PathUtils instead

bhapas committed:
Merge branch 'main' into cef_processor

Commit:
[CI] Auto commit changes from spotless

joegallo committed:
Merge branch 'main' into cef_processor

joegallo committed:
Clean up the exception handling

joegallo committed:
Use slightly more conventional test fixture idioms

joegallo committed:
Put the cef.txt files into a cef/ resources directory

joegallo committed:
Add a test that captures a deviation from the spec

joegallo committed:
Add a test for the backslash n edge case

joegallo committed:
The spec doesn't mention tabs. Handle them anymway.

joegallo committed:
Rename this to ignoreEmptyValues

joegallo committed:
Avoid an unnecessary try/catch for non-timestamps

joegallo committed:
Rewrite the parser

joegallo committed:
Merge branch 'main' into cef_processor

joegallo committed:
Remove trimTrailingCharacter changes

joegallo committed:
Merge branch 'main' into cef_processor

joegallo commented:
Caveats about microbenchmarking aside, here's this PR before my rewrite of the parser (that is, this is the regex based parser): 

```
              {
                "cef" : {
                  "type" : "cef",
                  "stats" : {
                    "count" : 108000,
                    "time_in_millis" : 2856,
                    "current" : 0,
                    "failed" : 20000
                  }
                }
              }
```

And here it is for the same workload after my rewrite:

```
              {
                "cef" : {
                  "type" : "cef",
                  "stats" : {
                    "count" : 108000,
                    "time_in_millis" : 1378,
                    "current" : 0,
                    "failed" : 20000
                  }
                }
              }
```

So at this point we're averaging about 12.7 microseconds per invocation. The workload for this was generated from the test fixtures cef messages, which explains why there are so many failures (since some of the fixture files contain illegal things to demonstrate that we fail on them).
